src/tests/unit/test_functions.py

```diff
@@ -248,6 +248,23 @@ def check(addr, expect_plen):
     check("54db:020a::a2cd:3c96:8ed8:7ea1:8540", 122)
 
 
+def test_ipaddrmask_make_subnet():
+    def s(addr, plen):
+        addrbin = helpers.ipaddr_to_bin(addr)
+        subnetbin = firewall.functions.ipaddrmask_make_subnet(addrbin, plen)
+        if addrbin == subnetbin:
+            assert addrbin is subnetbin
+        return helpers.ipaddr_from_bin(subnetbin)
+
+    assert s("192.168.0.0", 24) == "192.168.0.0"
+    assert s("192.168.0.1", 32) == "192.168.0.1"
+    assert s("192.168.0.3", 24) == "192.168.0.0"
+    assert s("192.168.0.3", 0) == "0.0.0.0"
+    assert s("aa::4:1", 128) == "aa::4:1"
+    assert s("aa::4:1", 127) == "aa::4:0"
+    assert s("aa::4:1", 0) == "::"
```

Comment:
`ipaddrmask_make_subnet()` and `ipaddrmask_from_plen()` are currently not used. They are a counter piece to `ipaddrmask_to_plen()`. I moved them to the unit tests.

How to implement `ipaddrmask_to_plen()` via `ipaddress`? Would that be

```python
try:
    ipaddress.IPv4Network(f'0.0.0.0/{addr}').prefixlen
except Exception:
    raise ValueError("hi")
```
that one raises an exception, if `{addr}` has holes (e.g. `ipaddress.IPv4Network(f'0.0.0.0/1.0.1.0').prefixlen`). That may or may not be desirable. `ipaddrmask_to_plen()` intentionally accepts holes, because that is what `checkIPnMask()` would have accepted. While the branch will tighten up some checks (and can break stuff that may have worked previously), it chooses not to break this one.

In general, calculating a prefix length for a subnet mask is fairly simple (aside the question, what to do about those holes). It doesn't require a python library, it requires unit tests that tie down the behavior we want. firewalld's job is to convert configurations (about IP addresses) around, it needs a strong understanding of what it exactly does. And it has its own Idiosyncrasies, like you mentioned, the [] brackets for IPv6 addresses.